flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManager.java

```diff
@@ -21,519 +21,897 @@
+		this.slotRequestTimeout = Preconditions.checkNotNull(slotRequestTimeout);
+		this.taskManagerTimeout = Preconditions.checkNotNull(taskManagerTimeout);
+
+		slots = new HashMap<>(16);
+		freeSlots = new LinkedHashMap<>(16);
+		taskManagerRegistrations = new HashMap<>(4);
+		fulfilledSlotRequests = new HashMap<>(16);
+		pendingSlotRequests = new HashMap<>(16);
+
+		leaderId = null;
+		resourceManagerActions = null;
+		started = false;
 	}
 
-	// ------------------------------------------------------------------------
-	//  slot managements
-	// ------------------------------------------------------------------------
+	// ---------------------------------------------------------------------------------------------
+	// Component lifecycle methods
+	// ---------------------------------------------------------------------------------------------
 
 	/**
-	 * Request a slot with requirements, we may either fulfill the request or pending it. Trigger container
-	 * allocation if we don't have enough resource. If we have free slot which can match the request, record
-	 * this allocation and forward the request to TaskManager through ResourceManager (we want this done by
-	 * RPC's main thread to avoid race condition).
+	 * Starts the slot manager with the given leader id and resource manager actions.
 	 *
-	 * @param request The detailed request of the slot
-	 * @return RMSlotRequestRegistered The confirmation message to be send to the caller
+	 * @param newLeaderId to use for communication with the task managers
+	 * @param newResourceManagerActions to use for resource (de-)allocations
+	 */
+	public void start(UUID newLeaderId, Executor newMainThreadExecutor, ResourceManagerActions newResourceManagerActions) {
+		leaderId = Preconditions.checkNotNull(newLeaderId);
+		mainThreadExecutor = Preconditions.checkNotNull(newMainThreadExecutor);
+		resourceManagerActions = Preconditions.checkNotNull(newResourceManagerActions);
+
+		started = true;
+	}
+
+	/**
+	 * Suspends the component. This clears the internal state of the slot manager.
 	 */
-	public RMSlotRequestRegistered requestSlot(final SlotRequest request) {
-		final AllocationID allocationId = request.getAllocationId();
-		if (isRequestDuplicated(request)) {
-			LOG.warn("Duplicated slot request, AllocationID:{}", allocationId);
-			return new RMSlotRequestRegistered(allocationId);
+	public void suspend() {
+		for (PendingSlotRequest pendingSlotRequest : pendingSlotRequests.values()) {
+			cancelPendingSlotRequest(pendingSlotRequest);
 		}
 
-		// try to fulfil the request with current free slots
-		final ResourceSlot slot = chooseSlotToUse(request, freeSlots);
-		if (slot != null) {
-			LOG.info("Assigning SlotID({}) to AllocationID({}), JobID:{}", slot.getSlotId(),
-				allocationId, request.getJobId());
+		pendingSlotRequests.clear();
+
+		HashSet<InstanceID> registeredTaskManagers = new HashSet<>(taskManagerRegistrations.keySet());
```

Comment:
If the copy of the collection is just for concurrent iteration, I would make it an `ArrayList`. Builds faster and iterates faster.